app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.kt

```diff
@@ -16,127 +16,133 @@
 
 package com.example.android.architecture.blueprints.todoapp.addedittask
 
-import androidx.lifecycle.LiveData
-import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.SavedStateHandle
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import com.example.android.architecture.blueprints.todoapp.R
+import com.example.android.architecture.blueprints.todoapp.TodoDestinationsArgs
 import com.example.android.architecture.blueprints.todoapp.data.Result.Success
 import com.example.android.architecture.blueprints.todoapp.data.Task
 import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.launch
 
+/**
+ * UiState for the Add/Edit screen
+ */
+data class AddEditTaskUiState(
+    val title: String = "",
+    val description: String = "",
+    val isTaskCompleted: Boolean = false,
+    val isLoading: Boolean = false,
+    val userMessage: Int? = null,
+    val isTaskSaved: Boolean = false
+)
+
 /**
  * ViewModel for the Add/Edit screen.
  */
 class AddEditTaskViewModel(
-    private val tasksRepository: TasksRepository
+    private val tasksRepository: TasksRepository,
+    savedStateHandle: SavedStateHandle
 ) : ViewModel() {
 
-    // Two-way databinding, exposing MutableLiveData
-    val title = MutableLiveData<String>()
-
-    // Two-way databinding, exposing MutableLiveData
-    val description = MutableLiveData<String>()
-
-    private val _dataLoading = MutableLiveData<Boolean>()
-    val dataLoading: LiveData<Boolean> = _dataLoading
-
-    private val _snackbarText = MutableLiveData<Int?>()
-    val snackbarText: LiveData<Int?> = _snackbarText
-
-    private val _taskUpdatedEvent = MutableLiveData(false)
-    val taskUpdatedEvent: LiveData<Boolean> = _taskUpdatedEvent
-
-    private var taskId: String? = null
-
-    private var isNewTask: Boolean = false
-
-    private var isDataLoaded = false
-
-    private var taskCompleted = false
-
-    fun start(taskId: String?) {
-        if (_dataLoading.value == true) {
-            return
-        }
-
-        this.taskId = taskId
-        if (taskId == null) {
-            // No need to populate, it's a new task
-            isNewTask = true
-            return
-        }
-        if (isDataLoaded) {
-            // No need to populate, already have data.
-            return
-        }
+    private val taskId: String? = savedStateHandle.get(TodoDestinationsArgs.TASK_ID_ARG)
 
-        isNewTask = false
-        _dataLoading.value = true
+    private val _uiState = MutableStateFlow(AddEditTaskUiState())
```

Comment:
😍

Reply:
You could get the taskId from the handle and use it in a flow {} that is combined with a userMessage flow to make the uiState. Why not?

Reply:
This ViewModel is a bit trickier to make fully functional because the source of truth of the current editable Task is the ViewModel, not the data layer.

Methods like `updateTitle` or `updateDescription` need to mutate the UI state directly, that's why we need a hold of the mutable UiState. The data layer load data is only valid for the first UiState exposed, not later.

Reply:
Going to add some comments about this in the code. Thanks!

Reply:
Oh right, sgtm